feat: Add llhls preloadSegment, add parts to segments, and fix byterange for parts/preloadHints

brandonocasey opened the pull request:
This does a few things:
1. Adds preloadSegment. Which is the segment that is going to be added to the playlist, but we only have parts and preload hints for right now.
2. Associates parts with segments
3. Better byterange parsing for bytes

brandonocasey commented on src/parse-stream.js:
```diff
@@ -5,6 +5,21 @@ import Stream from '@videojs/vhs-utils/es/stream.js';
 
 const TAB = String.fromCharCode(0x09);
 
+const parseByterange = function(byterangeString) {
```
Moved all byterange parsing into here as we do this for `map`, `byterange`, and not `part`

brandonocasey commented on src/parse-stream.js:
```diff
@@ -535,6 +543,11 @@ export default class ParseStream extends Stream {
           if (event.attributes.hasOwnProperty(key)) {
             event.attributes[key] = parseInt(event.attributes[key], 10);
+
+            const subkey = key === 'BYTERANGE-LENGTH' ? 'length' : 'offset';
```
This one is a bit different as we have attributes for both `offest` and `length`

brandonocasey commented on src/parser.js:
```diff
@@ -92,6 +92,7 @@ export default class Parser extends Stream {
     let currentMap;
     // if specified, the active decryption key
     let key;
+    let hasParts = false;
```
Added this to keep renditionReport warnings firing if we get a rendition report before any parts.

brandonocasey commented on src/parser.js:
```diff
@@ -115,6 +116,29 @@ export default class Parser extends Stream {
+        currentUri.timeline = currentTimeline;
+      }
+
+      this.manifest.preloadSegment = currentUri;
```
build preload segment if we get `end` without getting an actual segment uri.

brandonocasey commented on src/parser.js:
```diff
@@ -115,6 +116,29 @@ export default class Parser extends Stream {
     // previous segment
     let lastByterangeEnd = 0;
+    // keep track of the last seen part's byte range end.
+    let lastPartByterangeEnd = 0;
```
keep track of the current part byterange

gesinger commented on src/parser.js:
```diff
@@ -501,10 +560,10 @@ export default class Parser extends Stream {
 
                 this.trigger('warn', {
-                  message: `#EXT-X-PRELOAD-HINT #${index} lacks required attribute(s): ${missingAttributes.join(', ')}`
+                  message: `#EXT-X-PRELOAD-HINT #${index} for segment #${segmentIndex} lacks required attribute(s): ${missingAttributes.join(', ')}`
```
It might also be worth warning (because only `SHOULD NOT`) when we encounter two EXT-X-PRELOAD-HINT tags with the same TYPE. From https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-08#section-6.2.6

>  The Server SHOULD NOT add more than one EXT-X-PRELOAD-HINT tag with
   the same TYPE to a Playlist.

gesinger commented on test/fixtures/integration/llhls-byte-range.js:
```diff
@@ -0,0 +1,260 @@
+      {
+        'BYTERANGE-LENGTH': 5000,
+        'BYTERANGE-START': 8355216,
+        'TYPE': 'PART',
```
Is this valid due to https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-08#section-6.2.6 ?

> The Server SHOULD NOT add more than one EXT-X-PRELOAD-HINT tag with
the same TYPE to a Playlist.

(I suppose valid but not recommended, but may be worth identifying if so in the playlist name)

gesinger commented on test/fixtures/integration/llhls-byte-range.js:
```diff
@@ -0,0 +1,260 @@
+    {
+      duration: 10,
+      timeline: 0,
+      uri: 'hls_450k_video.ts'
```
Should this have a byterange?

gesinger commented on test/fixtures/integration/llhls-byte-range.js:
```diff
@@ -0,0 +1,260 @@
+          URI: 'hls_450k_video.ts',
+          byterange: {
+            length: 22444,
+            offset: 7673596
```
Is it meant for these offsets to be out of order?

gesinger commented on test/fixtures/integration/llhls-byte-range.m3u8:
```diff
@@ -0,0 +1,65 @@
+#EXTINF:1.4167,
+#EXT-X-BYTERANGE:44556@8353216
+hls_450k_video.ts
+#EXT-X-PRELOAD-HINT:TYPE=PART,URI="filePart273.3.mp4",BYTERANGE-LENGTH=2000
```
Same comment as above re: multiple EXT-X-PRELOAD-HINTs of the same type

gesinger commented on test/fixtures/integration/llhls-delta-byte-range.js:
```diff
@@ -0,0 +1,161 @@
+          URI: 'hls_450k_video.ts',
+          byterange: {
+            length: 22444,
+            offset: 8096148
```
Same comments as per above parsed result

gesinger commented on test/fixtures/integration/llhls-delta-byte-range.m3u8:
```diff
@@ -0,0 +1,30 @@
+hls_450k_video.ts
+#EXT-X-PART:DURATION=0.33334,URI="hls_450k_video.ts",BYTERANGE=22444
+#EXT-X-PRELOAD-HINT:TYPE=PART,URI="filePart273.3.mp4",BYTERANGE-LENGTH=2000
+#EXT-X-PRELOAD-HINT:TYPE=PART,URI="filePart273.4.mp4",BYTERANGE-LENGTH=5000,BYTERANGE-START=8377660
```
I guess this is causing the out of order byte ranges (and the start used in the other playlist)

brandonocasey commented:
* Update preload-hints so that it warns on duplicate `TYPE` preloadHints. Added a test for it too.
* Added a comment to byterange parsing regex
* Fixed issues with the test data regarding preload-hints of the same type and out-of-order byteranges.

gkatsev commented on src/parse-stream.js:
```diff
@@ -472,15 +485,12 @@ export default class ParseStream extends Stream {
-          if (offset) {
-            event.byterange.offset = parseInt(offset, 10);
-          }
+          event.attributes.byterange = {};
+          Object.keys(result).forEach(function(k) {
+            event.attributes.byterange[k] = result[k];
+          });
```
why not set `byterange` to be directly the response from `parseByterange`?

gkatsev commented on src/parse-stream.js:
```diff
@@ -263,15 +279,12 @@ export default class ParseStream extends Stream {
-            if (offset) {
-              event.byterange.offset = parseInt(offset, 10);
-            }
+            Object.keys(result).forEach(function(k) {
+              event.byterange[k] = result[k];
+            });
```
why not set byterange to be directly the response from parseByterange?

gkatsev commented on src/parse-stream.js:
```diff
@@ -222,18 +239,17 @@ export default class ParseStream extends Stream {
-        if (match[2]) {
-          event.offset = parseInt(match[2], 10);
-        }
+        const result = parseByterange(match[1]);
+
+        Object.keys(result).forEach(function(k) {
+          event[k] = result[k];
+        });
```
What about setting `event` to be the response from `parseByterange` and then add `type` and `tagType` to that object?

brandonocasey commented:
cool call on the byterange stuff @gkatsev
